### PR TITLE
Support EVAL_RO and EVALSHA_RO

### PR DESCRIPTION
**Pull request by Avital-Fine**

### CODE REVIEW

File: src/main/java/redis/clients/jedis/commands/ScriptingKeyCommands.java [hunk 1]
```diff
@@ -10,9 +10,13 @@
 
   Object eval(String script, List<String> keys, List<String> args);
 
+  Object evalReadonly(String script, List<String> keys, List<String> args);
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/ScriptingKeyCommands.java:13]: add javadoc.

File: src/main/java/redis/clients/jedis/commands/ScriptingKeyCommands.java [hunk 2]
```diff
@@ -10,9 +10,13 @@
...
+
   Object evalsha(String sha1);
 
   Object evalsha(String sha1, int keyCount, String... params);
 
   Object evalsha(String sha1, List<String> keys, List<String> args);
+
+  Object evalshaReadonly(String sha1, List<String> keys, List<String> args);
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/ScriptingKeyCommands.java:21]: add javadoc.

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 3]
```diff
@@ -2455,6 +2455,14 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
         BuilderFactory.ENCODED_OBJECT);
   }
 
+  public final CommandObject<Object> evalReadonly(String script, List<String> keys, List<String> args) {
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/CommandObjects.java:2458]: Why not set keys to `String[]`

**Review by zeekling — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/CommandObjects.java (hunk 3 above)
Comment (Avital-Fine) [reply src/main/java/redis/clients/jedis/CommandObjects.java:2458]: Was trying to be consistent with the `eval` functions

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Instead of `new ArrayList<...>(){}`, use `Collections.emptyList()`.

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 4]
```diff
@@ -2455,6 +2455,14 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
...
+    String[] keysArray = keys.toArray(new String[keys.size()]);
+    String[] argsArray = args.toArray(new String[args.size()]);
+    return new CommandObject<>(commandArguments(EVAL_RO).add(script).add(keysArray.length)
+            .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
+            BuilderFactory.ENCODED_OBJECT);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:2463]:
```suggestion
    return new CommandObject<>(commandArguments(EVAL_RO).add(script).add(keysArray.length)
        .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
        BuilderFactory.ENCODED_OBJECT);
```

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 5]
```diff
@@ -2477,6 +2485,14 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
         BuilderFactory.RAW_OBJECT);
   }
 
+  public final CommandObject<Object> evalReadonly(byte[] script, List<byte[]> keys, List<byte[]> args) {
+    byte[][] keysArray = keys.toArray(new byte[keys.size()][]);
+    byte[][] argsArray = args.toArray(new byte[args.size()][]);
+    return new CommandObject<>(commandArguments(EVAL_RO).add(script).add(keysArray.length)
+            .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
+            BuilderFactory.RAW_OBJECT);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:2493]:
```suggestion
    return new CommandObject<>(commandArguments(EVAL_RO).add(script).add(keysArray.length)
        .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
        BuilderFactory.RAW_OBJECT);
```

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 6]
```diff
@@ -2499,6 +2515,14 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
         BuilderFactory.ENCODED_OBJECT);
   }
 
+  public final CommandObject<Object> evalshaReadonly(String sha1, List<String> keys, List<String> args) {
+    String[] keysArray = keys.toArray(new String[keys.size()]);
+    String[] argsArray = args.toArray(new String[args.size()]);
+    return new CommandObject<>(commandArguments(EVALSHA_RO).add(sha1).add(keysArray.length)
+            .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
+            BuilderFactory.ENCODED_OBJECT);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:2523]:
```suggestion
    return new CommandObject<>(commandArguments(EVALSHA_RO).add(sha1).add(keysArray.length)
        .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
        BuilderFactory.ENCODED_OBJECT);
```

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 7]
```diff
@@ -2521,6 +2545,14 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
         BuilderFactory.RAW_OBJECT);
   }
 
+  public final CommandObject<Object> evalshaReadonly(byte[] sha1, List<byte[]> keys, List<byte[]> args) {
+    byte[][] keysArray = keys.toArray(new byte[keys.size()][]);
+    byte[][] argsArray = args.toArray(new byte[args.size()][]);
+    return new CommandObject<>(commandArguments(EVALSHA_RO).add(sha1).add(keysArray.length)
+            .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
+            BuilderFactory.RAW_OBJECT);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:2553]:
```suggestion
    return new CommandObject<>(commandArguments(EVALSHA_RO).add(sha1).add(keysArray.length)
        .keys((Object[]) keysArray).addObjects((Object[]) argsArray),
        BuilderFactory.RAW_OBJECT);
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Comment by sazzad16** [comment]
@Avital-Fine Please resolve conflicts.

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 8]
```diff
@@ -235,7 +235,8 @@ public static Object read(final RedisInputStream is) {
     GEORADIUSBYMEMBER, GEORADIUSBYMEMBER_RO, BITFIELD, HSTRLEN, TOUCH, SWAPDB, MEMORY, XADD, XLEN,
     XDEL, XTRIM, XRANGE, XREVRANGE, XREAD, XACK, XGROUP, XREADGROUP, XPENDING, XCLAIM, XAUTOCLAIM,
     XINFO, BITFIELD_RO, LPOS, ZMSCORE, BZPOPMIN, BZPOPMAX, BLMOVE, LMOVE, COPY, ROLE, FAILOVER,
-    STRALGO, GEOSEARCH, GEOSEARCHSTORE, LOLWUT, REPLICAOF, ZRANGESTORE, SINTERCARD, SORT_RO;
+    STRALGO, GEOSEARCH, GEOSEARCHSTORE, LOLWUT, REPLICAOF, ZRANGESTORE, SINTERCARD, SORT_RO, EVAL_RO,
+    EVALSHA_RO;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java:239]:
```suggestion
    STRALGO, GEOSEARCH, GEOSEARCHSTORE, LOLWUT, REPLICAOF, ZRANGESTORE, SINTERCARD, SORT_RO,
    EVAL_RO, EVALSHA_RO;
```

**Review by sazzad16 — DISMISSED** [review]

**Review by dengliming — APPROVED** [review]

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**